### PR TITLE
Adapt Protonet for ProtoMAML

### PR DESCRIPTION
**Pull request by ChristophHoenes**
adapted Protonet class with flexible MLPClassifier
some fixes including an important bugfix in MLPClassifier
new ProtoMAML model class with functions required in meta-train

### CODE REVIEW

File: util.py [hunk 1]
```diff
@@ -109,6 +109,19 @@ def make_dataloader(input_ids, labels, attention_masks, batch_size=16, shuffle=T
                             )
     return dataloader
 
+def split_episode(batch, ratio=.5):
+    batch_size = batch[0].size()[0]
+    support_size = round(batch_size*ratio)
+
+    support_idx_pos = np.random.choice(np.arange(0, batch_size, 2), int(np.ceil(support_size/2)), replace=False)
+    support_idx_neg = np.random.choice(np.arange(1, batch_size, 2), int(np.floor(support_size/2)), replace=False)
```
Comment (baleato) [inline util.py]: Can we assume that the task sampler will always give alternating examples (positive/negative)?
Otherwise we can look at the labels in batch[1] and split their indexes by label.

File: tasks.py [hunk 2]
```diff
@@ -59,6 +59,19 @@ def sample_next_task(self):
         else:
             return np.random.choice(len(self.task_iters), p=self.task_probs)
 
+    def get_task_batch(self, task_index):
```
Comment (baleato) [inline tasks.py]: I think this function is no longer used.

File: util.py (hunk 1 above)
Comment (ChristophHoenes) [reply util.py]: With the Episodic sampler this is the case. However in the case with three or more labels this approach would be flawed (not guaranteed to yield equal label distribution within support/query set although approximately true)
But as Phillip mentioned we should make the support and query set two completely separate data sets anyway (otherwise we will get previously seen examples in the query set)
If we make these changes then this function won't be needed anymore

File: tasks.py (hunk 2 above)
Comment (ChristophHoenes) [reply tasks.py]: True.

File: util.py (hunk 1 above)
Comment (baleato) [reply util.py]: Cool, let's wait for those changes then.

**Review by baleato — APPROVED** [review]
